cdap-watchdog/src/main/java/io/cdap/cdap/metrics/collect/AggregatedMetricsEmitter.java

```diff
@@ -33,38 +34,64 @@
   private final String name;
   // metric value
   private final AtomicLong value;
-  // specifies if the metric type is gauge or counter
-  private final AtomicBoolean gaugeUsed;
+
+  private volatile MetricType metricType = MetricType.COUNTER;
```

Comment:
That is how the original code was. If emit is called at the wrong time, the metric will remain a counter. A counter with value 0 is being ignored https://github.com/cdapio/cdap/blob/28f948fe1dc675f0e9263632c8b5c8f11ec90ad8/cdap-watchdog/src/main/java/io/cdap/cdap/metrics/collect/AggregatedMetricsCollectionService.java#L190.

Reply:
Actually, I discovered a race condition even in the existing implementation of Gauge. For a gauge metric, once the value is updated here https://github.com/cdapio/cdap/blob/39b6e5504edc8a7d9403eddc9d57fb22a4ba83a7/cdap-watchdog/src/main/java/io/cdap/cdap/metrics/collect/AggregatedMetricsEmitter.java#L63, if emit is called before the next line when gaugeUsed is set, it will be published as a counter metric instead of a Gauge metric.

To keep things simple, I have made all methods use synchronized. There are no Atomics or volatile.